virttest/migration.py

```diff
@@ -0,0 +1,326 @@
+                                                    args=(vm, desturi, options,
+                                                          ignore_status, virsh_opt,
+                                                          extra_opts))
+                migration_thread.start()
+                eclipse_time = 0
+                stime = int(time.time())
+                if func:
+                    # Execute command once the migration is started
+                    migrate_start_state = args.get("migrate_start_state", "paused")
+                    if self.wait_for_migration_start(vm, state=migrate_start_state, uri=desturi):
+                        logging.info("Migration started for %s", vm.name)
+                        if func == process.run:
+                            try:
+                                func(args['func_params'], shell=args['shell'])
+                            except KeyError:
+                                func(args['func_params'])
+                        elif func == virsh.migrate_postcopy:
+                            func(vm.name, uri=srcuri, debug=True)
+                        else:
+                            func(args['func_params'])
+                    else:
+                        logging.error("Migration failed to start for %s",
+                                      vm.name)
+                eclipse_time = int(time.time()) - stime
+                logging.debug("start_time:%d, eclipse_time:%d", stime, eclipse_time)
+                if eclipse_time < thread_timeout:
+                    migration_thread.join(thread_timeout - eclipse_time)
+                if migration_thread.isAlive():
+                    logging.error("Migrate %s timeout.", migration_thread)
+                    self.RET_LOCK.acquire()
+                    self.RET_MIGRATION = False
+                    self.RET_LOCK.release()
+        elif migration_type == "cross":
+            # Migrate a vm to remote first,
+            # then migrate another to remote with the first vm back
+            vm_remote = vms.pop()
+            self.thread_func_migration(vm_remote, desturi)
+            for vm in vms:
+                thread1 = threading.Thread(target=self.thread_func_migration,
+                                           args=(vm_remote, srcuri, options))
+                thread2 = threading.Thread(target=self.thread_func_migration,
+                                           args=(vm, desturi, options))
+                thread1.start()
+                thread2.start()
+                thread1.join(thread_timeout)
+                thread2.join(thread_timeout)
+                vm_remote = vm
+                if thread1.isAlive() or thread1.isAlive():
+                    logging.error("Cross migrate timeout.")
+                    self.RET_LOCK.acquire()
+                    self.RET_MIGRATION = False
+                    self.RET_LOCK.release()
+            # Add popped vm back to list
+            vms.append(vm_remote)
```

Comment:
vm_remote gets reassigned in the iteration vm_remote=vm, so when comes out of loop, it vm_remote will point to last item in the list after popped, It is good to preserve the `vm_remote` with different name